### ISSUE TITLE
Use a sound naming strategy when generating identifiers

### ISSUE THREAD

**Issue by g-r-a-n-t** [ethereum/fe#477]
### What is wrong?

We need to make sure names generated by the compiler can't collide with user-defined names or other names generated by the compiler.

### How can it be fixed

Add well-documented naming modules to the codegen libraries and use them when generating identifiers.

**Comment by g-r-a-n-t** [comment]
closing #377

**Comment by g-r-a-n-t** [comment]
Copying @sbillig thoughts from https://github.com/ethereum/fe/pull/472#discussion_r663161327 here:

This will presumably allow for naming collisions when we support nested tuples and structs. Doesn't necessarily have to be addressed now, but we need standard method for collision-avoidance everywhere code is generated (as we discussed yesterday). In the lowering pass, I added a trailing underscore to the tuple name, but maybe we should just add an incrementing number to every generated identifier (eg add a `$N` suffix, which I think @Y-Nak suggested), rather than trying to differentiate between different nestings by adding a character that stands in for the parentheses.
eg
`(A, (B, C), D)` -> `abi_decode_component_tuple_A_tuple_B_C_D$0`
`(A, (B, C, D))` -> `abi_decode_component_tuple_A_tuple_B_C_D$1`

**Comment by g-r-a-n-t** [comment]
> but maybe we should just add an incrementing number to every generated identifier

I'm not opposed to doing this, but I think we still want to avoid ambiguous type names in Yul, as the [number one](https://docs.soliditylang.org/en/latest/yul.html#motivation-and-high-level-description) design goal is to be readable.

**Comment by sbillig** [comment]
> > but maybe we should just add an incrementing number to every generated identifier
> 
> I'm not opposed to doing this, but I think we still want to avoid ambiguous type names in Yul, as the [number one](https://docs.soliditylang.org/en/latest/yul.html#motivation-and-high-level-description) design goal is to be readable.

Yeah, I agree. I think we should attempt to generate useful names that don't collide, but I'm imagining new potential for naming collisions cropping up every time we add a new generated name construction, and it'll get harder to think through all the possible collision cases as we add more and more generation. Eg say we need to generate an identifier for a generic struct like `Foo<u8, u16>`, and we use `Foo_u8_16_`, then some fuzzer comes along and creates a generic struct called `tuple`, so we think of a way to avoid that particular collision, then the fuzzer finds some other weird combination, ...

So, I'm just suggesting that it would be nice to have a simple guarantee that there can't possibly be a collision.
